Validate multiplex images

ngreenwald opened the pull request:
This PR adds more thorough validation of the images before they are run through any models to produce informative error messages for users.

ngreenwald commented:
Build is currently erroring due to install on 2.7? I didn't change anything in requirements.txt, not sure why that would be

willgraf commented on redis_consumer/consumers/multiplex_consumer.py:
```diff
@@ -71,6 +71,33 @@ def _consume(self, redis_hash):
             # TODO: tiffs expand the last axis, is that a problem here?
             image = utils.get_image(fname)
 
+        # validate correct shape of image
+        if len(image.shape) > 3:
+            raise ValueError('An image of shape {} was supplied, but the '
+                             'multiplex model expects of images of shape'
+                             '[height, widths, 2]'.format(image.shape))
+        elif len(image.shape) < 3:
+            # TODO: Once we can pass warning messages to user, we can treat this as nuclear image
+            raise ValueError('A 2D image of shape {} was supplied, but the '
+                             'multiplex model expects images of shape'
+                             '[height, width, 2]')
+        else:
```
It seems like a lot of overlapping conditions here. Are you leaving it separated out for future handling of all these conditions? Can any of the conditions be joined, just for brevity?

ngreenwald replied:
Yeah, the reason I didn't just do

```
if len(image.shape != 3):
    raise ValueError()
```

Is for the TODO on line 80. Once we can pass warnings, we can treat data that's length 2 as if it's missing a channels axis and only has the nuclear channel, and reformat it accordingly

willgraf commented on redis_consumer/consumers/multiplex_consumer.py:
```diff
@@ -71,6 +71,33 @@ def _consume(self, redis_hash):
+                raise ValueError('An image of shape {} was supplied, but the '
+                                 'multiplex model expects of images of shape'
+                                 '[height, widths, 2]'.format(image.shape))
+            elif image.shape[0] == 2:
+                # TODO: once we can pass warning messages tell user we switched to channels last
+                image = np.rollaxis(image, 0, 3)
+
+            elif image.shape[2] == 2:
+                pass
```
Do we need to warn them about channels_first? Maybe we can just remember what the data came in as, and roll the axis again at the end for the output (if the output even has a channel axis).

The example data is all channels_first because ImageJ won't properly load channels_last data.

ngreenwald replied:
I actually ran into the opposite problem: When I supplied the data as channels_first deepcell.org failed, and it only worked with channels last. Looks like the tests also use channels_last: https://github.com/vanvalenlab/kiosk-redis-consumer/blob/0c6a4e3908780226c2cc7d6082e05eb6f8291126/redis_consumer/consumers/multiplex_consumer_test.py#L87



The predictions don't have a channels axis, they're just 2D tiffs. I assume that even once we start returning multiple predictions (for nuclear and whole_cell, for example), these would be separate 2D tiffs.

willgraf replied:
I think deepcell.org is deployed with [this branch of the consumer,](https://github.com/vanvalenlab/kiosk-redis-consumer/tree/multiplex-channels-first) which does handle channels_first, though maybe there's another problem going on.

Either way, I'm not sure we will need to tell the user at all if we switch the axis.

willgraf commented on redis_consumer/consumers/multiplex_consumer.py:
```diff
@@ -71,6 +71,33 @@ def _consume(self, redis_hash):
+                                 '[height, widths, 2]'.format(image.shape))
+            elif image.shape[0] == 2:
+                # TODO: once we can pass warning messages tell user we switched to channels last
+                image = np.rollaxis(image, 0, 3)
```
This line is untested.

willgraf commented on redis_consumer/consumers/multiplex_consumer.py:
```diff
@@ -71,6 +71,33 @@ def _consume(self, redis_hash):
+            elif image.shape[2] == 2:
+                pass
+            else:
+                raise ValueError('An image of shape {} was supplied, but the '
```
This line is untested.

willgraf commented:
I believe that the python 2.7 build problems have been fixed by #126. Please merge master into this branch to make sure.

ngreenwald commented:
I addressed the test coverage issues, and removed the warning about switching channels axis. Also removed the np.squeeze command that isn't necessary anymore.

willgraf commented:
Are you positive the squeeze is not required anymore? I think `get_image` adds a channel dimension to the image no matter what, as tiff files are not expected to have a channel image, but 2D images will have one.

ngreenwald commented:
Updated the tests to reflect what get_image actually does. Also exported newest version of the model and put it in the bucket, update the environment variable
